HousingFinanceInterimApi.Tests/V1/TestHelpers/ClearTable.cs

```diff
@@ -0,0 +1,71 @@
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.EntityFrameworkCore.Metadata;
+
+namespace HousingFinanceInterimApi.Tests.V1.TestHelpers
+{
+    public static class ClearTable
+    {
+        /// <summary>
+        /// Clear the table of the specified entity.
+        /// </summary>
+        public static void Clear(DbContext context, System.Type entitySysType)
+        {
+            var entityType = context.Model.FindEntityType(entitySysType);
+            context.Database.ExecuteSqlRaw($"DELETE FROM {entityType.GetSchema()}.{entityType.GetTableName()}");
+            context.SaveChanges();
+        }
+
+        /// <summary>
+        /// Clears the tables in the correct order based on foreign key dependencies.
+        /// </summary>
+        /// <param name="context"></param>
+        /// <param name="entityTypes"></param>
+        public static void ClearTables(this DbContext context, List<System.Type> entityTypes)
+        {
+            var sortedEntities = GetEntitiesInDependencyOrder(context, entityTypes);
+            foreach (var entity in sortedEntities)
+            {
+                Clear(context, entity.ClrType);
+            }
+        }
+
+        private static List<IEntityType> GetEntitiesInDependencyOrder(DbContext context, List<System.Type> entityTypes)
+        {
+            var entities = context.Model.GetEntityTypes()
+            .Where(e => entityTypes.Contains(e.ClrType))
+            .ToList();
+
+            var sortedEntities = new List<IEntityType>();
+            var visitedEntities = new HashSet<IEntityType>();
+
+            void Visit(IEntityType entity)
+            {
+                if (!visitedEntities.Contains(entity))
+                {
+                    visitedEntities.Add(entity);
+
+                    foreach (var foreignKey in entity.GetForeignKeys())
+                    {
+                        if (entityTypes.Contains(foreignKey.PrincipalEntityType.ClrType))
+                        {
+                            Visit(foreignKey.PrincipalEntityType);
+                        }
+                    }
+                    sortedEntities.Add(entity);
+                }
+            }
```

Comment:
Nested methods are quite hard to read as it breaks the 'I do' flow by inserting "here's a definition of what I can do" flow.

Reply:
tbh this whole thing was a copy-paste because I figured it was an already solved problem. I ran it and it worked so I kept it.

Here's an alternative without the nested method and recursion. Tests still pass so it's still working:

https://github.com/LBHackney-IT/housing-finance-interim-api/pull/149/commits/53bfbdcebecb125a2531d871be257c45e0c41c37